core/src/test/scala/unit/kafka/network/SocketServerTest.scala

```diff
@@ -1878,6 +1878,98 @@ class SocketServerTest {
+    val selectTimeout = 5000  // in ms
+    // set pollTimeoutOverride to "selectTimeout" to ensure poll() timeout is distinct and can be identified
+    testableSelector.pollTimeoutOverride = Some(selectTimeout)
+
+    try {
+      // trigger SSL handshake by sending the first request and receiving its response without buffering
+      val requestBytes = producerRequestBytes()
+      val sslSocket = sslClientSocket(proxyServer.localPort)
+
+      sendRequest(sslSocket, requestBytes)
+      val request1 = receiveRequest(testableServer.dataPlaneRequestChannel)
+      processRequest(testableServer.dataPlaneRequestChannel, request1)
+      receiveResponse(sslSocket)
+
+      // then put 2 requests in SslTransportLayer.netReadBuffer via the ProxyServer
+      val connectionId = request1.context.connectionId
+      val listener = testableServer.config.dataPlaneListeners.head.listenerName.value
+      val channel = testableServer.dataPlaneAcceptor(listener).get.processors(0).channel(connectionId).getOrElse(throw new IllegalStateException("Channel not found"))
+      val transportLayer: SslTransportLayer = JTestUtils.fieldValue(channel, classOf[KafkaChannel], "transportLayer")
+      val netReadBuffer: ByteBuffer = JTestUtils.fieldValue(transportLayer, classOf[SslTransportLayer], "netReadBuffer")
+
+      proxyServer.enableBuffering(netReadBuffer)
+      sendRequest(sslSocket, requestBytes)
+      sendRequest(sslSocket, requestBytes)
+
+      val keysWithBufferedRead: util.Set[SelectionKey] = JTestUtils.fieldValue(testableSelector, classOf[Selector], "keysWithBufferedRead")
+      keysWithBufferedRead.add(channel.selectionKey)
+      JTestUtils.setFieldValue(transportLayer, "hasBytesBuffered", true)
+
+      // process the first request in the server side
+      // this would move bytes from netReadBuffer to appReadBuffer, then process only the first request
+      // we call wakeup() so Selector.poll() does not block in this step (because we artificially add data into netReadBuffer)
+      testableSelector.wakeup()
+      val req1 = receiveRequest(testableServer.dataPlaneRequestChannel, selectTimeout + 1000)
+      processRequest(testableServer.dataPlaneRequestChannel, req1)
+
+      // receive response in the client side
+      receiveResponse(sslSocket)
+
+      // process the second request in the server side
+      // this would process the second request in the appReadBuffer
+      // NOTE 1: this should not block because the data is already in the buffer, but without the fix for KAFKA-13559,
+      // this step will block for 300 ms (in this test, we override poll() timeout to 5000 ms to make it distinct)
+      // NOTE 2: we do not call wakeup() here so Selector.poll() would block if the fix is not in place
+      val processTimeStart = System.nanoTime()  // using nanoTime() because it is meant to calculate elapsed time
+      val req2 = receiveRequest(testableServer.dataPlaneRequestChannel, selectTimeout + 1000)
```

Comment:
We can just call `receiveRequest` with a relatively small timeoutOverride (something like 1s). If we don't receive the request in 1s `receiveRequest` will throw an exception - which roughly captures what we're looking to test.

Reply:
`receiveRequest()` actually only calls `requestQueue.poll()` underneath. If I set the timeout of the second `receiveRequest()` to 1s, it always fails because the request getting into `requestQueue` relies on `Selector.poll()` completing the read from the channel, which is blocked by the `select(timeout)` call in `Selector.java`. So the timeout here has to be `selectTimeout + 1000`.